### ISSUE TITLE
Replace the outdated Unicode handling

### ISSUE THREAD

**Issue by ajtribick** [CelestiaProject/Celestia#1519]
Celestia contains its own implementation of UTF-8 (which permits things like 5- and 6-byte sequences), plus rather outdated Unicode normalization tables. There are also routines for case-insensitive comparisons, and numeric/datetime formatting.

I think it would be better to take a dependency on ICU and use its facilities for these things rather than maintaining our own implementation. This would also make it easier for some of the data-related components to do things like culture-specific collation, e.g. in the search panel.

I've been doing some initial experiments here so would be willing to take this on.

**Comment by 375gnu** [comment]
ICU is huuuge.

**Comment by ajtribick** [comment]
Yeah that is the downside. Another approach that might work is to provide a function like `setUnicodeHandler()` and offload the relevant functions to whatever front-end is in use, e.g. Qt's implementation on that front-end, the Windows API, etc. Then Celestia's code would only need to be able to handle the encoding details. The disadvantage is a lot of these seem to prefer UTF-16 (and that includes ICU in the case that the C++ APIs aren't available, e.g. the one supplied on Android or Windows 10/11).

Alternatively icu4x looks promising but their numeric formatting support is quite basic right now and the data generation is a little awkward. Plus it's not available on vcpkg and would introduce a requirement for a Rust compiler.

**Comment by 375gnu** [comment]
We even don't support RTL text rendering, so I believe we can live with our current unicode implementation.

**Comment by levinli303** [comment]
we can make it a cmake option that is by default disabled, especially given that integrating icu could help us implement RTL.

**Comment by ajtribick** [comment]
I think we do need to do this because at the moment we don't even casefold non-ASCII characters, so even if you can input æ, you have to input it as Æ to match Ægir.

From what I can see, we should be able to use WinAPI functions to replicate most if not all of the ICU functionality we would need on Windows. Unfortunately Qt's functions don't quite get there as they don't appear to expose enough low-level details of the collation algorithm to build a reasonable culture-aware autocomplete.

So I'm thinking that we could get away with defaulting icu to OFF on Windows in favour of a WinAPI-based implementation and ON for Linux (size of icu is less relevant as this is likely to be already installed as a shared library supplied by the distro). Not sure what the situation is on Apple platforms.

**Comment by levinli303** [comment]
On Apple platforms, ICU is bundled with the system however, the macOS SDK only gives a small set of icu header files and it is advised to not use it. you can use CFString https://developer.apple.com/documentation/corefoundation/cfstring which is purely in C.

But given that probably we are not going to distribute celestia-qt as a bundled application on Mac, it should be fine to use icu as well, because they can just install icu from homebrew. For mobile distributions I will build ICU for these.
